### PR TITLE
Semi-automatic ordering of products from suppliers

### PR DESCRIPTION
**Pull request by kjagiello**
For every product one can define base stock level (the desired quantity). Then, when placing an order from a supplier, system will aid by suggesting which products should be ordered, based on the current demand. In case of Narlivs, the "aiding" process will be that the system will prefill the cart at Narlivs with the items that should be ordered.

### CODE REVIEW

File: src/shop/models.py [hunk 1]
```diff
@@ -87,6 +97,15 @@ class Meta:
         verbose_name_plural = _('supplier products')
         unique_together = ('supplier', 'sku',)
 
+    @property
+    def unit_price(self):
+        return self.price / self.qty_multiplier
```
Comment (flaeppe) [inline src/shop/models.py]: It might be unlikely, but couldn't the multiplier end up as `0` here? I didn't see any barriers in the admin model.

File: src/shop/suppliers/base.py [hunk 2]
```diff
@@ -32,3 +32,13 @@ def retrieve_product(self, sku):
         :type sku: str
         :rtype Union[SupplierProduct, None]
         """
+
+    @abstractmethod
+    def order_product(self, sku, qty):
+        """Places an order on product with given SKU.
```
Comment (flaeppe) [inline src/shop/suppliers/base.py]: What is `SKU`? Please write the full definition at least once before using the abbreviation

File: src/shop/tests/test_api.py [hunk 3]
```diff
@@ -27,9 +32,13 @@ def parse_delivery_report(self, report_path):
     def retrieve_product(self, sku):
         return SupplierProduct(
             name='Billys Original',
-            price=Decimal('9.25')
+            price=Decimal('9.25'),
+            units=1
         )
 
+    def order_product(self, sku, qty):
+        pass
```
Comment (flaeppe) [inline src/shop/tests/test_api.py]: Is this a drive-by test? Or maybe just forgotten?

File: src/shop/tests/test_api.py [hunk 4]
```diff
@@ -391,3 +404,80 @@ def test_update_quantity_prediction(self, predict_quantity_mock):
...
+            qty_multiplier=10,
+            units=1
+        )
+        factories.SupplierProductFactory(
+            product=product2,
+            supplier=supplier2,
+            price=1,
+            units=1
+        )
+        sp5 = api.order_from_supplier(product1.id, 48)
```
Comment (flaeppe) [inline src/shop/tests/test_api.py]: In my eyes, there is some black magic going on here, this should raise an exception.

File: src/shop/models.py (hunk 1 above)
Comment (kjagiello) [inline src/shop/models.py]: That's very true.

File: src/shop/api.py [hunk 5]
```diff
@@ -331,3 +348,70 @@ def update_out_of_stock_forecast(product_id):
...
+        try:
+            supplier_api.order_product(product.sku, minimum_qty(product))
+            return product
+        except SupplierAPIException:
+            # Log the error and try the next product.
+            log.warning('Failed to order product SKU %d from %s.',
+                        product.sku, supplier.internal_name)
+    else:
+        msg = 'Could not order {}.'.format(product.sku)
+        raise exceptions.APIException(msg)
```
Comment (kjagiello) [inline src/shop/api.py]: `else` won't be executed if you do a `break` in your loop. https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

I do not know however how I feel about translating internal exceptions. I know that I'm displaying their messages in some places (exceptions from the stock taking module) in the admin panel, but it feels kind of wrong. I'll think about it and get back to you.

File: src/shop/suppliers/base.py (hunk 2 above)
Comment (kjagiello) [inline src/shop/suppliers/base.py]: SKU is a well known abbreviation when you dealing with inventory management, but a clarification won't hurt anybody I guess!

File: src/shop/tests/test_api.py (hunk 3 above)
Comment (kjagiello) [inline src/shop/tests/test_api.py]: It's just something we need to do as long we keep `DummySupplierAPI` around. All the methods in `SupplierBase` are decorated with `@abstractmethod` and all the children classes need to implement all those methods, otherwise hell will break loose and Python won't run your code.

Comment (kjagiello) [inline src/shop/tests/test_api.py]: `order_product` is tested using another mocking method and we could probably do the same for `retrieve_product` and `parse_delivery_report`. Then, that dummy class could be removed entirely.

File: src/shop/tests/test_api.py (hunk 4 above)
Comment (kjagiello) [inline src/shop/tests/test_api.py]: Why an exception?

File: src/shop/api.py (hunk 5 above)
Comment (flaeppe) [inline src/shop/api.py]: Yeah, ignore my first section of the comment. I completely missed the return statement

File: src/shop/suppliers/base.py (hunk 2 above)
Comment (flaeppe) [inline src/shop/suppliers/base.py]: Well, that's a good justification, however it just saves readers trouble to write the full definition once at least.

File: src/shop/tests/test_api.py (hunk 4 above)
Comment (flaeppe) [inline src/shop/tests/test_api.py]: Because of my first comment. I have soon finished my coffee and hope the embarrassment goes away soon.

File: src/shop/tests/test_api.py (hunk 3 above)
Comment (flaeppe) [inline src/shop/tests/test_api.py]: I see, yeah maybe a small refactor could be in order.

File: src/shop/tests/test_api.py (hunk 4 above)
Comment (kjagiello) [inline src/shop/tests/test_api.py]: ☕️
